### PR TITLE
[SDK-618] - setUserLocation call added with all parameters of location.

### PR DESCRIPTION
**Pull request by ijunaid**
-- Fix warnings and remove deprecated calls from examples App.
-- Updated underlying android SDK to 21.11.0-RC4

### CODE REVIEW

File: lib/countly_flutter.dart [hunk 1]
```diff
@@ -1422,7 +1437,7 @@ class Countly {
     log('Calling "logExceptionEx":[$exceptionString] nonfatal:[$nonfatal]');
     stacktrace ??= StackTrace.current;
     final result = logException(
-        '${exceptionString}\n\n$stacktrace', nonfatal, segmentation);
+        '$exceptionString\n\n$stacktrace', nonfatal, segmentation);
```
Comment (ArtursKadikis) [inline lib/countly_flutter.dart:1440]: is this correct?

File: ios/Classes/CountlyFlutterPlugin.m [hunk 2]
```diff
@@ -359,7 +359,34 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+            CLLocationCoordinate2D location = kCLLocationCoordinate2DInvalid;
+            if(gpsCoordinates && [gpsCoordinates containsString:@","]){
+                @try{
+                    NSArray *locationArray = [gpsCoordinates componentsSeparatedByString:@","];
+                    NSString* latitudeString = [locationArray objectAtIndex:0];
+                    NSString* longitudeString = [locationArray objectAtIndex:1];
+
+                    double latitudeDouble = [latitudeString doubleValue];
+                    double longitudeDouble = [longitudeString doubleValue];
+                    location = (CLLocationCoordinate2D){latitudeDouble,longitudeDouble};
```
Comment (ArtursKadikis) [inline ios/Classes/CountlyFlutterPlugin.m:375]: What happens if you pass:
",10" or "10,"?

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
this PR has your server credentials.

File: ios/Classes/CountlyFlutterPlugin.m [hunk 3]
```diff
@@ -359,7 +359,34 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+         dispatch_async(dispatch_get_main_queue(), ^ {
+            NSDictionary* location = [command objectAtIndex:0];
+            NSString* gpsCoordinates =  location[@"gpsCoordinates"];
+            CLLocationCoordinate2D locationCoordinate = kCLLocationCoordinate2DInvalid;
+            if(gpsCoordinates && [gpsCoordinates containsString:@","]){
+                @try{
+                    NSArray *locationArray = [gpsCoordinates componentsSeparatedByString:@","];
+                    NSString* latitudeString = [locationArray objectAtIndex:0];
+                    NSString* longitudeString = [locationArray objectAtIndex:1];
+
```
Comment (ArtursKadikis) [inline ios/Classes/CountlyFlutterPlugin.m:372]: we need 2 warnings for each of the value

Comment (ArtursKadikis) [reply ios/Classes/CountlyFlutterPlugin.m:372]: add also a third warning in case array has more values

File: ios/Classes/CountlyFlutterPlugin.m [hunk 4]
```diff
@@ -359,7 +359,34 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
         result(@"setLocation!");
         });
 
-    }else if ([@"enableCrashReporting" isEqualToString:call.method]) {
+    }else if ([@"setUserLocation" isEqualToString:call.method]) {
+         dispatch_async(dispatch_get_main_queue(), ^ {
+            NSDictionary* location = [command objectAtIndex:0];
+            NSString* gpsCoordinates =  location[@"gpsCoordinates"];
+            CLLocationCoordinate2D locationCoordinate = kCLLocationCoordinate2DInvalid;
+            if(gpsCoordinates && [gpsCoordinates containsString:@","]){
```
Comment (ArtursKadikis) [inline ios/Classes/CountlyFlutterPlugin.m:367]: the lack of a comma should be communicated. Print error

File: ios/Classes/CountlyFlutterPlugin.m [hunk 5]
```diff
@@ -359,7 +359,34 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+                @try{
+                    NSArray *locationArray = [gpsCoordinates componentsSeparatedByString:@","];
+                    NSString* latitudeString = [locationArray objectAtIndex:0];
+                    NSString* longitudeString = [locationArray objectAtIndex:1];
+
+                    double latitudeDouble = [latitudeString doubleValue];
+                    double longitudeDouble = [longitudeString doubleValue];
```
Comment (ArtursKadikis) [inline ios/Classes/CountlyFlutterPlugin.m:374]: after this line. If any of the location values are 0, print warning of potential error and location string

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

File: ios/Classes/CountlyFlutterPlugin.m (hunk 3 above)
Comment (ArtursKadikis) [reply ios/Classes/CountlyFlutterPlugin.m:372]: skip the "empty string" warning

File: lib/countly_flutter.dart (hunk 1 above)
Comment (ijunaid) [reply lib/countly_flutter.dart:1440]: yes, it is correct.

File: example/ios/Flutter/Flutter.podspec [hunk 6]
```diff
@@ -11,7 +11,7 @@ Pod::Spec.new do |s|
   s.license          = { :type => 'MIT' }
   s.author           = { 'Flutter Dev Team' => 'flutter-dev@googlegroups.com' }
   s.source           = { :git => 'https://github.com/flutter/engine', :tag => s.version.to_s }
-  s.ios.deployment_target = '10.0'
+  s.ios.deployment_target = '9.0'
```
Comment (ArtursKadikis) [inline example/ios/Flutter/Flutter.podspec:14]: this should be reverted

File: CHANGELOG.md [hunk 7]
```diff
@@ -5,12 +5,13 @@
 * Added Appear and dismiss callback for nps/survey widgets
 * Added "recordIndirectAttribution" method
 * Added "recordDirectAttribution" method (Currently supporterd only for Android)
+* Added "setUserLocation" method to set user location
```
Comment (ArtursKadikis) [inline CHANGELOG.md:8]: Mention that the old call was deprecated

Comment (ArtursKadikis) [reply CHANGELOG.md:8]: and also mention the old initLocation call

Comment (ArtursKadikis) [reply CHANGELOG.md:8]: "Deprecated old init config methods. You should use the config object now. Those methods are: ..."

File: lib/countly_flutter.dart [hunk 8]
```diff
@@ -639,17 +641,30 @@ class Countly {
...
     }
-    List<String> args = [];
+    final String? result = await setUserLocation(gpsCoordinates: '$latitude,$longitude');
+    return result;
+  }
 
-    args.add(latitude);
-    args.add(longitude);
-    
+  /// Set user location
```
Comment (ArtursKadikis) [inline lib/countly_flutter.dart:648]: mention that all parameters are optional, but at least one has to be set

File: ios/Classes/CountlyFlutterPlugin.m [hunk 9]
```diff
@@ -359,7 +359,46 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
         result(@"setLocation!");
         });
 
-    }else if ([@"enableCrashReporting" isEqualToString:call.method]) {
+    }else if ([@"setUserLocation" isEqualToString:call.method]) {
+         dispatch_async(dispatch_get_main_queue(), ^ {
+            NSDictionary* location = [command objectAtIndex:0];
+             NSString* gpsCoordinates = location[@"gpsCoordinates"];
```
Comment (ArtursKadikis) [inline ios/Classes/CountlyFlutterPlugin.m:365]: empty space "typo"

File: ios/Classes/CountlyFlutterPlugin.m [hunk 10]
```diff
@@ -359,7 +359,46 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+
+                        double latitudeDouble = [latitudeString doubleValue];
+                        double longitudeDouble = [longitudeString doubleValue];
+                        if(latitudeDouble == 0 || longitudeDouble == 0) {
+                            COUNTLY_FLUTTER_LOG(@"Invalid location Coordinates: '%@', it should contains two comma seperated double values", gpsCoordinates);
+                        }
+                        locationCoordinate = (CLLocationCoordinate2D){latitudeDouble,longitudeDouble};
+                    }
+                    @catch(NSException *exception){
+                        COUNTLY_FLUTTER_LOG(@"Invalid location Coordinates: '%@'", gpsCoordinates);
```
Comment (ArtursKadikis) [inline ios/Classes/CountlyFlutterPlugin.m:385]: wrap the exception in "[ ]"

Comment (ArtursKadikis) [reply ios/Classes/CountlyFlutterPlugin.m:385]: remove empty space

File: ios/Classes/CountlyFlutterPlugin.m [hunk 11]
```diff
@@ -359,7 +359,46 @@ - (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
...
+                        if(locationArray.count > 2) {
+                            COUNTLY_FLUTTER_LOG(@"Invalid location Coordinates: '%@', it should contains only two comma seperated values", gpsCoordinates);
+                        }
+                        NSString* latitudeString = [locationArray objectAtIndex:0];
+                        NSString* longitudeString = [locationArray objectAtIndex:1];
+
+                        double latitudeDouble = [latitudeString doubleValue];
+                        double longitudeDouble = [longitudeString doubleValue];
+                        if(latitudeDouble == 0 || longitudeDouble == 0) {
+                            COUNTLY_FLUTTER_LOG(@"Invalid location Coordinates: '%@', it should contains two comma seperated double values", gpsCoordinates);
```
Comment (ArtursKadikis) [inline ios/Classes/CountlyFlutterPlugin.m:380]: this warning needs to be reworked

Comment (ArtursKadikis) [reply ios/Classes/CountlyFlutterPlugin.m:380]: warn about the 0 in the coordinates

Comment (ArtursKadikis) [reply ios/Classes/CountlyFlutterPlugin.m:380]: "One of the values parsed to a 0, double check that given coordinates are correct:[$coordinates]."

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]

**Comment by ArtursKadikis** [comment]
Tweak the init time data structure

**Review by ArtursKadikis — CHANGES_REQUESTED** [review]
Let's prefix the location fields/entries with "location".
Pull out the ios location value checking into a separate function.

**Pull request merged**